### ISSUE TITLE
Fails to build on macOS 10.15 Beta (19A471t)

### ISSUE THREAD

**Issue by KizzyCode** [sodiumoxide/sodiumoxide#341]
## Expected:
The crate builds

## What happens:
The crate fails to build with the following output:
```
   Compiling libc v0.2.58
   Compiling crc32fast v1.2.0
   Compiling byteorder v1.3.1
   Compiling cfg-if v0.1.9
   Compiling adler32 v1.0.3
   Compiling cc v1.0.37
   Compiling pkg-config v0.3.14
   Compiling serde v1.0.92
   Compiling filetime v0.2.6
   Compiling xattr v0.2.2
   Compiling libflate v0.1.23
   Compiling tar v0.4.26
   Compiling libsodium-sys v0.2.2 (/Users/keziah/Desktop/sodiumoxide/libsodium-sys)
error: failed to run custom build command for `libsodium-sys v0.2.2 (/Users/keziah/Desktop/sodiumoxide/libsodium-sys)`
process didn't exit successfully: `/Users/keziah/Desktop/sodiumoxide/target/release/build/libsodium-sys-4ccade80ee9f97d2/build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-env-changed=SODIUM_LIB_DIR
cargo:rerun-if-env-changed=SODIUM_SHARED
cargo:rerun-if-env-changed=SODIUM_USE_PKG_CONFIG
cargo:rerun-if-env-changed=SODIUM_DISABLE_PIE
OPT_LEVEL = Some("3")
TARGET = Some("x86_64-apple-darwin")
HOST = Some("x86_64-apple-darwin")
CC_x86_64-apple-darwin = None
CC_x86_64_apple_darwin = None
HOST_CC = None
CC = None
CFLAGS_x86_64-apple-darwin = None
CFLAGS_x86_64_apple_darwin = None
HOST_CFLAGS = None
CFLAGS = None
CRATE_CC_NO_DEFAULTS = None
DEBUG = Some("false")
CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2,sse3,ssse3")

--- stderr
thread 'main' panicked at 'Failed to run './configure': No such file or directory (os error 2)
', libsodium-sys/build.rs:370:13
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

warning: build failed, waiting for other jobs to finish...
error: build failed
```

## Steps to reproduce:
 0. Clone the repo
 1. Run `cargo build` or `cargo build --release`

## Workaround:
Use an installed libsodium version by setting `SODIUM_USE_PKG_CONFIG=1`

**Comment by kpp** [comment]
@KizzyCode You got no C compiler installed:

```
CC = None
```

Please install gcc or clang and try again

**Comment by KizzyCode** [comment]
`clang` is installed; `which clang` outputs `/usr/bin/clang`

To be clear: I've no problem building `libsodium` from source using `./configure`, `make` and it passes all tests pass if I run `make check`...

**Comment by kpp** [comment]
Try https://github.com/sodiumoxide/sodiumoxide#using-precompiled-library

**Comment by KizzyCode** [comment]
Yeah; this is what my workaround does (except that it uses pkg-config so that I don't have to specify the location manually).

I just wanted to open that issue so that it can be resolved before Catalina comes out finally.

**Comment by kpp** [comment]
Thanks for the issue. We will try to solve it

**Comment by agrinman** [comment]
Confirming this is an issue, I'm getting the same error as well

**Comment by kpp** [comment]
@agrinman can you create a user on your mac and give me an ssh access?

**Comment by nvesely** [comment]
Please ignore the request for SSH access.

Can you confirm libsodium is being downloaded and unarchived correctly? The error seems to be showing the that `configure` file within libsodium archive is not where our Rust build script expects it to be. Can you clone a fresh copy of the repository and show the output of `find . -iname '*configure*'` after an (ostensibly failed) build? Running build with `RUST_BACKTRACE=1` can't hurt either, though I don't know if we'll get anything more useful than we already have in this case.

**Comment by kpp** [comment]
> Please ignore the request for SSH access.

Why?

> Can you confirm libsodium is being downloaded and unarchived correctly? 

We don't download libsodium anymore.

Still you can try to debug it via github =)

**Comment by agrinman** [comment]
@nvesely I can confirm that configure is there and the download and unarchiving was successful. The strange part is that the Command still can’t seem to run configure. I added an ls command that runs ‘ls configure’ on the same source dir and it is successful. I’m thinking that it might be related to macos’s Security features around accessing files in the users docs directory or something to that effect, but can’t understand why it’s failing in this case only.

**Comment by KizzyCode** [comment]
Well, _in theory_ the security features shouldn't be a problem since I granted `Terminal.app` full disk access and the developer permission to run unsigned apps (which seems to work until now).
But since this is the first beta, chances are good that this is really due to a macOS bug which will be resolved in the next versions...
`./configure` is marked as executable?

**Comment by jedisct1** [comment]
In order to execute a command with a relative path, `.` now needs to be in the `PATH` environment variable.

**Comment by agrinman** [comment]
Not sure what's going on here:

```rust
use std::process::Command;

fn main() {
    let out = Command::new("./script").current_dir("/Users/Alex/Documents/projects/cmdtest/test")
        .output().expect("failed to exec command");
    eprintln!("output: {:?}", out);
}
```

Where there's a simple shell script located at `/test/script`. 

In case A:
```sh
echo "exec file test";
```

In case B:
```sh
#!/bin/sh
echo "exec file test";
```

When the test script is A, it succeeds as expected. When the shebang is included, as in B, it fails with the following error:

```
thread 'main' panicked at 'failed to exec command: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:997:5
```

If I change `./script` to `script`, case B doesn't panic but spits out a message about zsh being the new default shell:

```
output: Output { status: ExitStatus(ExitStatus(0)), stdout: "^D\u{8}\u{8}\r\nThe default interactive shell is now zsh.\r\nTo update your account to use zsh, please run `chsh -s /bin/zsh`.\r\nFor more details, please visit https://support.apple.com/kb/HT208050```

Any idea what's going on here? Might changing to zsh really fix this behavior?

**Comment by KizzyCode** [comment]
Additionally if you call `./configure` with it's absolute path instead of it's relative path, it works too 😳
To me this looks really weird too and I'd not be surprised if this is a macOS bug...

**Comment by KizzyCode** [comment]
@jedisct1 I was unable to modify the path variable to make the relative path working – how did you do it?

**Comment by jedisct1** [comment]
@KizzyCode 

```rust
configure_cmd.env("PATH", format!("{}:.", env::var("PATH").unwrap_or(String::new())));
```

I think this behavior is intentional since macOS now includes `.` in the default `PATH`.

**Comment by KizzyCode** [comment]
@jedisct1 
> I think this behavior is intentional since macOS now includes `.` in the default `PATH`

It does? On my fresh install, all `PATH`-variables contain the usual dirs but not `.`

**Comment by stv0g** [comment]
See: https://doc.rust-lang.org/std/process/struct.Command.html#platform-specific-behavior

It mentions that we should not rely on `Command::current_dir()` while using relative paths for the executable. My PR uses `fs::canonicalize()` as proposed by the docs.

**Comment by stv0g** [comment]
For some reason the compilation still fails because of [some bug in libsodium](https://github.com/jedisct1/libsodium/issues/848)

**Comment by eliaslevy** [comment]
FWIW, this issue appears fixed in Xcode 11.2 and the crate builds successfully with it. Note that if you had installed the command line tools, using sudo xcode-select -s /Applications/Xcode.app after updating Xcode is insufficient. You must remove the command line tools (sudo rm -rf /Library/Developer/CommandLineTools) or the build will still fail.

**Comment by kpp** [comment]
Resolved
